crypto/evp/evp_lib.c

```diff
@@ -242,6 +242,12 @@ int EVP_CIPHER_iv_length(const EVP_CIPHER *cipher)
 
 int EVP_CIPHER_CTX_iv_length(const EVP_CIPHER_CTX *ctx)
 {
+    int i;
+
+    if ((EVP_CIPHER_flags(ctx->cipher) & EVP_CIPH_CUSTOM_IV_LENGTH) != 0
+            && EVP_CIPHER_CTX_ctrl((EVP_CIPHER_CTX *)ctx, EVP_CTRL_GET_IVLEN,
+                                   0, &i) == 1)
+        return i;
     return ctx->cipher->iv_len;
```

Comment:
So, if the `EVP_CIPHER_CTX_ctrl()` returns with an *error* (i.e. 0), we don't care and return the legacy CIPHER IV length anyway.

Me, I'm thinking that this should happen:

If `EVP_CIPHER_CTX_ctrl()` returns `EVP_CTRL_RET_UNSUPPORTED`, then `ctx->cipher->iv_len` should be returned.
If `EVP_CIPHER_CTX_ctrl()` returns 0, then -1 should be returned (because 0 is a perfectly valid IV length, while -1 is not).
If `EVP_CIPHER_CTX_ctrl()` returns 1, then `i` should be returned.

There's a reason for the dance we do in other places, such as this:

https://github.com/openssl/openssl/blob/e2e5abe47af60260e9a4247597e64a4e9d266a7a/crypto/evp/evp_lib.c#L219-L228

(although, `EVP_CTRL_RET_UNSUPPORTED` is kinda off there, -1 only indicates an error at that stage)

Reply:
This dance you refer to is causing quite a few coverity issues currently - Since EVP_CIPHER_CTX_num() can now return -1 (and we dont check for negative any place where it is called - there ar elots of calls to this in the e_*.c functions)

Reply:
Sure, there are consequences to this.  We do need to deal with them.  Contrast that to basically ignoring errors and returning some value that may, or may not be accurate...  I'm not too keen on the latter.

Reply:
Either way,  it's a pattern with these value returning functions, and it's not even new.  See for example (from 1.1.1):

https://github.com/openssl/openssl/blob/77fa495f7f7cc2368c8c799e657e173a29a90a49/crypto/evp/evp_lib.c#L313-L320

Reply:
I think it if it has that flag set then it should either error or return the value.
Otherwise it returns the default value.

Reply:
So if `EVP_CIPHER_CTX_ctrl()` returns 0, we're just gonna pretend there was no error...  in that case, you need to clear the error queue

Reply:
Ah, I see that you're made changes, hold on...